File: src/timeline/js/directives/ruler.js
/**
 * @file
 * @brief Ruler directives (dragging playhead functionality, progress bars, tick marks, etc...)
 * @author Jonathan Thomas <jonathan@openshot.org>
 * @author Cody Parker <cody@yourcodepro.com>
 *
 * @section LICENSE
 *
 * Copyright (c) 2008-2018 OpenShot Studios, LLC
 * <http://www.openshotstudios.com/>. This file is part of
 * OpenShot Video Editor, an open-source project dedicated to
 * delivering high quality video editing and animation solutions to the
 * world. For more information visit <http://www.openshot.org/>.
 *
 * OpenShot Video Editor is free software: you can redistribute it
 * and/or modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * OpenShot Video Editor is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OpenShot Library.  If not, see <http://www.gnu.org/licenses/>.
 */


/*global App, timeline, secondsToTime, setSelections, setBoundingBox, moveBoundingBox, bounding_box */
// Variables for panning by middle click
var is_scrolling = false;
var starting_scrollbar = {x: 0, y: 0};
var starting_mouse_position = {x: 0, y: 0};

// Variables for scrolling control
var scroll_left_pixels = 0;


// This container allows for tracks to be scrolled (with synced ruler)
// and allows for panning of the timeline with the middle mouse button
App.directive("tlScrollableTracks", function () {
  return {
    restrict: "A",
    link: function (scope, element, attrs) {

      /**
       * if ctrl is held, scroll in or out.
       * Implimentation copied from zoomSlider zoomIn zoomOut
       */
      element.on("wheel",function (e) {
        if (e.ctrlKey) {
          e.preventDefault(); // Don't scroll like a browser
          var delta = e.originalEvent.deltaY * (e.shiftKey ? -1 : 1);
          if (delta > 0) { // Scroll down: Zoom out
            /*global timeline*/
            timeline.zoomOut();
          } else { // Scroll Up: Zoom in
            /*global timeline*/
            timeline.zoomIn();
          }
        }
        else if (e.shiftKey) {
          e.preventDefault();
          let current_scroll = $("#scrolling_tracks").scrollLeft();
          $("#scrolling_tracks").scrollLeft(current_scroll + e.originalEvent.deltaY);
        }
      });

      // Sync ruler to track scrolling
      element.on("scroll", function () {
        var scrollLeft = element.scrollLeft();
        var timelineWidth = scope.getTimelineWidth(0); // Full width of the timeline
        var maxScrollLeft = timelineWidth - element.width(); // Max horizontal scroll

        // Clamp to right edge
        element.scrollLeft(Math.min(scrollLeft, maxScrollLeft));

        // Sync the ruler and other components
        $("#track_controls").scrollTop(element.scrollTop());
        $("#scrolling_ruler, #progress_container").scrollLeft(scrollLeft);

        // Send scrollbar position to Qt if available
        if (scope.Qt) {
          // Create variables first and pass them as arguments
          var leftScrollbarEdge = scrollLeft / timelineWidth; // Use the full timeline width
          var rightScrollbarEdge = (scrollLeft + element.width()) / timelineWidth; // Use the full timeline width

          // Pass the variables as a JavaScript array (interpreted as a PyQt list)
          timeline.ScrollbarChanged([leftScrollbarEdge, rightScrollbarEdge, timelineWidth, element.width()]);
        }

        // Update scrollLeft in scope
        scope.$apply(() => scope.scrollLeft = scrollLeft);
      });

      // Pans the timeline (on middle mouse click and drag)
      element.on("mousemove", function (e) {
        if (is_scrolling) {
          var difference = {x: starting_mouse_position.x - e.pageX, y: starting_mouse_position.y - e.pageY};
          var newPos = {
            x: Math.max(0, Math.min(starting_scrollbar.x + difference.x, scope.getTimelineWidth(0) - element.width())),
            y: Math.max(0, Math.min(starting_scrollbar.y + difference.y, $("#scrolling_tracks")[0].scrollHeight - element.height()))
          };

          // Scroll the tracks div
          element.scrollLeft(newPos.x).scrollTop(newPos.y);
        }
      });

      // Remove move cursor (i.e. dragging has stopped)
      element.on("mouseup", function (e) {
        element.removeClass("drag_cursor");
      });

    }
  };
});

// Track scrolling mode on body tag... allows for capture of released middle mouse button
App.directive("tlBody", function () {
  return {
    link: function (scope, element, attrs) {

      element.on("mouseup", function (e) {
        if (e.which === 2) { // middle button
          is_scrolling = false;
          element.removeClass("drag_cursor");
        }
      });

      // Stop scrolling if mouse leaves timeline
      element.on("mouseleave", function (e) {
        is_scrolling = false;
        element.removeClass("drag_cursor");
      })

      // Initialize panning when middle mouse is clicked
      element.on("mousedown", function (e) {
        if (e.which === 2) { // middle button
          e.preventDefault();
          is_scrolling = true;
          starting_scrollbar = {x: $("#scrolling_tracks").scrollLeft(), y: $("#scrolling_tracks").scrollTop()};
          starting_mouse_position = {x: e.pageX, y: e.pageY};
          element.addClass("drag_cursor");
        }
      });

    }
  };
});

// The HTML5 canvas ruler
App.directive("tlRuler", function ($timeout) {
  return {
    restrict: "A",
    link: function (scope, element, attrs) {
      var isDragging = false;

      // Start dragging when mousedown on the ruler
      element.on("mousedown", function (e) {
        // Set bounding box for the playhead position
        setBoundingBox(scope, $("#playhead"), "playhead");
        isDragging = true;

        if (scope.Qt) {
            // Disable caching thread during scrubbing
            timeline.DisableCacheThread();
        }

        // Get playhead position
        var playhead_left = e.pageX - element.offset().left;
        var playhead_seconds = snapToFPSGridTime(scope, pixelToTime(scope, playhead_left));
        playhead_seconds = Math.min(Math.max(0.0, playhead_seconds), scope.project.duration);
        var playhead_snapped_target = playhead_seconds * scope.pixelsPerSecond;

        // Immediately preview frame (don't wait for animated playhead)
        scope.previewFrame(playhead_seconds);

        if (playhead_seconds == scope.project.playhead_position) {
          // No animation (playhead didn't move)
          return;
        }

        // Animate to new position (and then update scope)
        scope.playhead_animating = true;
        $(".playhead-line").animate({left: playhead_snapped_target}, 150);
        $(".playhead-top").animate({left: playhead_snapped_target}, 150, function () {
          // Update playhead
          scope.movePlayhead(playhead_seconds);

          // Animation complete.
          scope.$apply(function () {
            scope.playhead_animating = false;
          });
        });
      });

      // Global mousemove listener
      $(document).on("mousemove", function (e) {
        if (isDragging && e.which === 1 && !scope.playhead_animating && !scope.getDragging()) { // left button is held
          // Calculate the playhead bounding box movement
          let cursor_position = e.pageX - $("#ruler").offset().left;
          let new_position = cursor_position;
          if (e.shiftKey) {
            // Only apply playhead snapping when SHIFT is pressed
            let results = moveBoundingBox(scope, bounding_box.left, bounding_box.top,
              cursor_position - bounding_box.left, cursor_position - bounding_box.top,
              cursor_position, cursor_position, "playhead");

            // Update position to snapping position
            new_position = results.position.left;
          }

          // Move playhead
          let playhead_seconds = new_position / scope.pixelsPerSecond;
          playhead_seconds = Math.min(Math.max(0.0, playhead_seconds), scope.project.duration);
          scope.movePlayhead(playhead_seconds);
          scope.previewFrame(playhead_seconds);
        }
      });

      // Global mouseup listener to stop dragging
      $(document).on("mouseup", function (e) {
        if (isDragging) {
          isDragging = false;

          if (scope.Qt) {
            // Enable caching thread after scrubbing
            timeline.EnableCacheThread();
          }
        }
      });

      /**
       * Draw frame precision alternating banding on each track (when zoomed in)
       */
      function updateTrackBackground() {
          const fps = scope.project.fps.num / scope.project.fps.den;
          const pixelsPerFrame = scope.pixelsPerSecond / fps;
          const fpt = framesPerTick(scope.pixelsPerSecond, scope.project.fps.num, scope.project.fps.den);

          // Calculate the time start and end positions
          let timeStart = scope.scrollLeft / scope.pixelsPerSecond;
          let timeEnd = (scope.scrollLeft + $("body").width()) / scope.pixelsPerSecond;

          // Align the start time based on frames per tick
          timeStart -= fpt > fps ? (timeStart % (fpt / Math.round(fps))) : (timeStart % 2);
          timeEnd -= timeEnd % 1 - 1;

          const startFrame = timeStart * Math.round(fps);
          const endFrame = timeEnd * Math.round(fps);

          if (fpt <= 2.0) {
              $('.track').each(function() {
                  const $bandingContainer = $(this).find('.banding-overlay').empty();
                  const trackHeight = $(this).innerHeight();
                  let frame = startFrame;
                  let isDarkBand = true;

                  while (frame <= endFrame) {
                      const pos = (frame / fps) * scope.pixelsPerSecond;

                      if (fpt === 1.0) {
                          // When fpt == 1, alternate entire ticks as dark or transparent
                          if (isDarkBand) {
                              $bandingContainer.append($('<div></div>').css({
                                  'position': 'absolute',
                                  'left': `${pos}px`,
                                  'width': `${pixelsPerFrame}px`,  // Full frame width for one frame
                                  'height': '100%',
                                  'background-color': 'rgba(0, 0, 0, 0.1)'
                              }));
                          }
                          isDarkBand = !isDarkBand;  // Alternate bands
                      } else if (fpt === 2.0) {
                          // When fpt == 2, create two alternating dark bands per tick
                          const bandWidth = pixelsPerFrame * fpt;
                          $bandingContainer.append($('<div></div>').css({
                              'position': 'absolute',
                              'left': `${pos}px`,
                              'width': `${bandWidth / 2}px`,  // Dark band width is half of the tick
                              'height': '100%',
                              'background-color': 'rgba(0, 0, 0, 0.1)'
                          }));
                      }
                      frame += fpt;  // Move to the next frame
                  }

                  // Set container CSS for banding overlay
                  $bandingContainer.css({
                      'position': 'absolute',
                      'top': '0',
                      'left': '0',
                      'height': `${trackHeight}px`,
                      'width': '100%',
                      'z-index': '1',
                      'pointer-events': 'none'
                  });
              });
          } else {
              // Clear bands when not zoomed in enough
              $('.track').find('.banding-overlay').empty();
          }
      }

      /**
       * Draw times on the ruler
       * Always starts on a second
       * Draws to the right edge of the screen
       */
      function drawTimes() {
        // Delete old tick marks
        ruler = $('#ruler');
        $("#ruler span").remove();

        let startPos = scope.scrollLeft;
        let endPos = scope.scrollLeft + $("body").width();
        let fpt = framesPerTick(scope.pixelsPerSecond, scope.project.fps.num ,scope.project.fps.den);
        let fps = scope.project.fps.num / scope.project.fps.den;
        let time = [ startPos / scope.pixelsPerSecond, endPos / scope.pixelsPerSecond];

        if (fpt > fps) {
          // Make sure seconds don't change when scrolling right and left
          time[0] -= time[0]%(fpt/Math.round(fps));
        }
        else {
          time[0] -= time[0]%2;
        }
        time[1] -= time[1]%1 - 1;

        let startFrame = time[0] * Math.round(fps);
        let endFrame = time[1] * Math.round(fps);

        let frame = startFrame;
        while ( frame <= endFrame){
          let t = frame / fps;
          let pos = t * scope.pixelsPerSecond;
          let tickSpan = $('<span style="left:'+pos+'px;"></span>');
          tickSpan.addClass("tick_mark");

          if ((frame) % (fpt * 2) === 0) {
            // Alternating long marks with times marked
            let timeSpan = $('<span style="left:'+pos+'px;"></span>');
            timeSpan.addClass("ruler_time");
            let timeText = secondsToTime(t, scope.project.fps.num, scope.project.fps.den);
            timeSpan[0].innerText = timeText['hour'] + ':' +
              timeText['min'] + ':' +
              timeText['sec'];
            if (fpt < Math.round(fps)) {
              timeSpan[0].innerText += ',' + timeText['frame'];
            }
            tickSpan[0].style['height'] = '20px';
            ruler.append(timeSpan);
          }
          ruler.append(tickSpan);
          frame += fpt;
        }

        // Add alternating banding to indicate frame boundaries (when zoomed in)
        updateTrackBackground();
      }

      scope.$watch("project.scale + project.duration + scrollLeft + element.width()", function (val) {
        $timeout(function () {
          drawTimes();
          return;
        } , 0);
      });

    }

  };
});

File: src/windows/views/timeline_backend/paint/ruler.py
"""
 @file
 @brief Painter for the timeline ruler and header.
 @author Jonathan Thomas <jonathan@openshot.org>

 @section LICENSE

 Copyright (c) 2008-2025 OpenShot Studios, LLC
 (http://www.openshotstudios.com). This file is part of
 OpenShot Video Editor (http://www.openshot.org), an open-source project
 dedicated to delivering high quality video editing and animation solutions
 to the world.

 OpenShot Video Editor is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OpenShot Video Editor is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with OpenShot Library.  If not, see <http://www.gnu.org/licenses/>.
 """

from qt_api import QPointF, QRectF, Qt
from qt_api import (
    QBrush,
    QColor,
    QFont,
    QFontMetrics,
    QLinearGradient,
    QPainter,
    QPen,
)
import math

from classes.app import get_app
from classes.time_parts import secondsToTime

from .base import BasePainter


def _text_width(metrics, text):
    if hasattr(metrics, "horizontalAdvance"):
        return metrics.horizontalAdvance(text)
    return metrics.width(text)


class RulerPainter(BasePainter):
    def update_theme(self):
        self.bg = self.w.theme.ruler.background
        self.bg2 = self.w.theme.ruler.background2
        self.name_bg = (
            self.w.theme.ruler_name_background
            if self.w.theme.ruler_name_background.isValid()
            else self.w.theme.track.name_background
        )
        self.name_bg2 = (
            self.w.theme.ruler_name_background2
            if self.w.theme.ruler_name_background2.isValid()
            else self.name_bg
        )
        self.tick_pen = QPen(self.w.theme.ruler.border_color)
        self.tick_pen.setCosmetic(True)
        minor_col = QColor(self.w.theme.ruler.border_color)
        if minor_col.isValid():
            minor_col.setAlpha(int(minor_col.alpha() * 0.5))
        self.minor_tick_pen = QPen(minor_col)
        self.minor_tick_pen.setCosmetic(True)
        self.text_pen = QPen(self.w.theme.ruler.font_color)
        self.tick_font = QFont()
        if self.w.theme.ruler.font_size:
            self.tick_font.setPointSize(self.w.theme.ruler.font_size)
        self.play_font = QFont()
        if self.w.theme.ruler_time_font_size:
            self.play_font.setPointSize(self.w.theme.ruler_time_font_size)
        fam = str(getattr(self.w.theme.ruler, "font_family", "") or "")
        if fam:
            self.tick_font.setFamily(fam)
            self.play_font.setFamily(fam)
        self.label_top = self.w.theme.ruler_label_top
        self.pad_left = self.w.theme.ruler_time_pad_left
        self.pad_top = self.w.theme.ruler_time_pad_top
        self._last_playhead_label = ""

    def _current_playhead_label(self):
        proj = get_app().project
        fps_info = proj.get("fps")
        fps_float = float(fps_info.get("num", 24)) / float(fps_info.get("den", 1) or 1)
        frame_seconds = 0.0
        if fps_float:
            frame_seconds = max(
                0.0, (max(1, self.w.current_frame) - 1) / fps_float
            )
        tt = secondsToTime(
            frame_seconds,
            fps_info["num"],
            fps_info["den"],
        )
        return f"{tt['hour']}:{tt['min']}:{tt['sec']},{tt['frame']}"

    def _draw_time_panel(self, painter: QPainter, label: str = None):
        left_rect = QRectF(0, 0, self.w.track_name_width, self.w.ruler_height)
        if left_rect.width() <= 0 or left_rect.height() <= 0:
            return left_rect
        if self.name_bg2 != self.name_bg:
            grad = QLinearGradient(QPointF(left_rect.topLeft()), QPointF(left_rect.bottomLeft()))
            grad.setColorAt(0, self.name_bg)
            grad.setColorAt(1, self.name_bg2)
            painter.fillRect(left_rect, QBrush(grad))
        else:
            painter.fillRect(left_rect, self.name_bg)

        if label is None:
            label = self._current_playhead_label()
        if label:
            painter.setPen(self.text_pen)
            painter.setFont(self.play_font)
            painter.drawText(
                left_rect.adjusted(self.pad_left, self.pad_top, -2, -2),
                Qt.AlignLeft | Qt.AlignTop,
                label,
            )
            painter.setPen(self.tick_pen)
        return left_rect

    def _nice_frame_intervals(self, fps):
        fps_frames = max(1, int(round(fps)))
        intervals = set()

        for frames in (1, 2, 3, 5, 10, 15):
            if frames < fps_frames:
                intervals.add(frames)

        for fraction in (0.2, 0.25, 1.0 / 3.0, 0.5, 1.0):
            raw_frames = fps_frames * fraction
            frames = int(round(raw_frames))
            if frames >= 1 and abs(raw_frames - frames) < 1e-6:
                intervals.add(frames)

        for seconds in (2, 3, 5, 10, 15, 30, 60, 120, 300, 600, 900, 1800, 3600):
            intervals.add(fps_frames * seconds)

        return sorted(intervals)

    def _frames_per_tick(self, pps, fps):
        if pps <= 0 or fps <= 0:
            return 1

        min_spacing = 44.0
        for frames in self._nice_frame_intervals(fps):
            if (frames / fps) * pps >= min_spacing:
                return frames

        frames = self._nice_frame_intervals(fps)[-1]
        while (frames / fps) * pps < min_spacing:
            frames *= 2
        return frames

    def paint(self, painter: QPainter):
        proj = get_app().project
        duration = proj.get("duration")
        fps_info = proj.get("fps")
        fps_float = float(fps_info.get("num", 24)) / float(fps_info.get("den", 1) or 1)
        pps = self.w.pixels_per_second
        width = max(1, self.w.width() - self.w.track_name_width)

        rect = QRectF(self.w.track_name_width, 0, width, self.w.ruler_height)
        if self.bg2.isValid() and self.bg != self.bg2:
            grad = QLinearGradient(QPointF(rect.topLeft()), QPointF(rect.bottomLeft()))
            grad.setColorAt(0, self.bg)
            grad.setColorAt(1, self.bg2)
            painter.fillRect(rect, QBrush(grad))
        elif self.bg.isValid():
            painter.fillRect(rect, self.bg)
        play_lbl = self._current_playhead_label()
        self._last_playhead_label = play_lbl
        self._draw_time_panel(painter, play_lbl)
        base_y = self.w.ruler_height
        tick_metrics = QFontMetrics(self.tick_font)
        label_top = max(0, self.label_top - 2)
        long_ht = base_y - (label_top + tick_metrics.height()) - 2
        short_ht = long_ht / 2
        painter.setPen(self.tick_pen)

        offset_px = getattr(self.w, "h_scroll_offset", 0.0)
        if pps <= 0:
            return
        visible_px = width
        start_seconds = offset_px / pps
        end_seconds = (offset_px + visible_px) / pps
        fpt = self._frames_per_tick(pps, fps_float)
        if fpt <= 0:
            return
        total_frames = int(duration * fps_float)
        start_frame = int(math.floor((start_seconds * fps_float) / fpt) * fpt)
        start_frame = max(0, start_frame)
        end_frame = int(math.ceil((end_seconds * fps_float) / fpt) * fpt)
        end_frame = min(total_frames, end_frame + fpt)
        frame = start_frame
        while frame <= end_frame:
            t = frame / fps_float
            x = self.w.track_name_width + t * pps - offset_px
            is_major = (frame % (fpt * 2) == 0)
            ht = long_ht if is_major else short_ht

            if x >= self.w.track_name_width - 2 and x <= self.w.track_name_width + visible_px + 2:
                painter.setPen(self.tick_pen if is_major else self.minor_tick_pen)
                painter.drawLine(QPointF(x, base_y), QPointF(x, base_y - ht))

            if frame % (fpt * 2) == 0 and (
                x + 1.0 >= self.w.track_name_width and x <= self.w.track_name_width + visible_px
            ):
                tt = secondsToTime(t, fps_info["num"], fps_info["den"])
                if frame == 0:
                    lbl = f"{int(tt['min'])}:{tt['sec']}"
                    text_w = _text_width(tick_metrics, lbl)
                    text_rect = QRectF(
                        x + 2,
                        label_top,
                        text_w,
                        tick_metrics.height(),
                    )
                    align = Qt.AlignLeft | Qt.AlignTop
                else:
                    lbl = f"{tt['hour']}:{tt['min']}:{tt['sec']}"
                    if fpt < round(fps_float):
                        lbl += f",{tt['frame']}"
                    text_w = _text_width(tick_metrics, lbl)
                    text_rect = QRectF(
                        x - text_w / 2,
                        label_top,
                        text_w,
                        tick_metrics.height(),
                    )
                    align = Qt.AlignCenter | Qt.AlignTop
                painter.setPen(self.text_pen)
                painter.setFont(self.tick_font)
                painter.drawText(text_rect, align, lbl)
                painter.setPen(self.tick_pen)
            frame += fpt

    def paint_overlay(self, painter: QPainter):
        self._draw_time_panel(painter, getattr(self, "_last_playhead_label", None))
